src/coins.cpp

```diff
@@ -158,7 +158,7 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
                 // Otherwise we will need to create it in the parent
                 // and move the data up and mark it as dirty
                 CCoinsCacheEntry& entry = cacheCoins[it->first];
-                entry.coin = std::move(it->second.coin);
+                entry.coin = erase ? std::move(it->second.coin) : it->second.coin;
```

Comment:
Is the `std::move` gain really worth the added code complexity here?

Reply:
This has no effect, I think. All std::move does is cast to an rvalue reference. The ternary operator merges it back to the same type with the other branch.

Reply:
It is indeed required. Try recompiling without this change and running the coins tests (`./src/test/test_bitcoin -t coins_tests`).

The std::move is necessary to erase the contents of the map being written to the parent.

Reply:
@jamesob The `std::move` here has no effect; it is equivalent to `entry.coin = it->second.coin`. If you want to conditionally move, you need two separate statements.

Reply:
Did you actually try recompiling with your suggested change and running the tests? The result on my platform (`gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0`) is failure.

Reply:
@jamesob It sounds like you're relying on the side effect of the moved-from object being cleared or something, but AFAIK it is in fact left in an undefined state...

Probably should just copy and explicitly clear.

Reply:
It seems I'm wrong; your code is correct.

Apparently what happens in this case is that a temporary variable is created, which it->second.coin is either copied or moved to (depending on `erase`), and then that temporary variable is move-assigned to `entry.coin`. This has the desired effect, though there is possibly an unnecessary (but cheap) move.

If you write it as

```c++
if (erase) {
    entry.coin = std::move(it->second.coin);
} else {
    entry.coin = it->second.coin;
}
```

I think you get the same semantics, but without the additional move.

Reply:
@luke-jr Well, the only member with move semantics is CScript, which is a prevector, and move assigning is well defined (it's in our own code), namely swapping.

Reply:
Abstractly, it is undefined, and `std::move` AFAIK does not guarantee move assignment, only allows it?

Reply:
@luke-jr `std::move` just casts to an rvalue reference, nothing else. It's up to the involved classes to define what to do when assigned with or constructed from an rvalue.

The standard also defines what standard library classes do in this case: moved-from objects are in a valid but unspecified state (which means that you're only allowed to run operations on them that have no preconditions; e.g. empty() on them or assigning to them is ok).

However, no standard library classes are involved here, so it doesn't apply.

I agree with you insofar that if the desired effect is clearing, the code should just explicitly clear (potentially after a move, if that's more efficient). But it isn't UB to rely on the rvalue semantics of a class we defined ourselves.

Reply:
@Sjors please see above - we do rely on the `std::move` to clear the entry from the child cache, it isn't just a performance optimization. If you remove the `move`, the coins unittests will fail.

That said I agree it seems tricky to rely upon this behavior implicitly based upon classes we've define ourselves, but I haven't tested the performance implications of adding `mapCoins.erase(it)`. Given that I'm essentially leaving the behavior as it was before this change in the case of the `move`, and we have unittests that will catch a change in behavior here, I don't feel the need to add an explicit `erase()` call. I have however clarified the phrasing with an explicit if-else as @sipa suggests.

Reply:
> If you remove the move, the coins unittests will fail.

Odd, I tried that. Will try again.

Reply:
> Odd, I tried that. Will try again.

Perhaps you made the change in the relevant commit, but then didn't apply subsequent commits including the new tests? Just a guess.

Reply:
FWIW I did try this before acking. I checked out this branch and changed the ternary to just use the copy assignment and the tests errored out. Will try again with updated commits.

Reply:
@jnewbery has found that I was wrong - the std::move invocation isn't necessary for correctness, it's just an optimization. Thanks for pointing that out @Sjors. Will remove the new comment.

Reply:
Since this has caused so much confusion, I think it'd be better to update the comment than remove it entirely. Perhaps something like:

```
// We'll erase to clear this entry so we can move to the parent cache instead of copying

Reply:
Since this has caused so much confusion, I think it'd be better to update the comment than remove it entirely. Perhaps something like:

```
// We'll erase this entry from the map in the for increase expression.
// Move the coin to the parent cache instead of copying as an optimization.

Reply:
Pfew, I'm not insane :-) Thanks for adding the comment.